test/e2e_node/restart_test.go

```diff
@@ -88,34 +89,48 @@ var _ = framework.KubeDescribe("Restart [Serial] [Slow] [Disruptive]", func() {
 				// startTimeout fit on the node and the node is now saturated.
 				runningPods := waitForPods(f, podCount, startTimeout)
 				if len(runningPods) < minPods {
-					framework.Failf("Failed to start %d pods, cannot test that restarting docker doesn't leak IPs", minPods)
+					framework.Failf("Failed to start %d pods, cannot test that restarting container runtime doesn't leak IPs", minPods)
 				}
 
 				for i := 0; i < restartCount; i += 1 {
-					By(fmt.Sprintf("Restarting Docker Daemon iteration %d", i))
-
-					// TODO: Find a uniform way to deal with systemctl/initctl/service operations. #34494
-					if stdout, err := exec.Command("sudo", "systemctl", "restart", "docker").CombinedOutput(); err != nil {
-						framework.Logf("Failed to trigger docker restart with systemd/systemctl: %v, stdout: %q", err, string(stdout))
-						if stdout, err = exec.Command("sudo", "service", "docker", "restart").CombinedOutput(); err != nil {
-							framework.Failf("Failed to trigger docker restart with upstart/service: %v, stdout: %q", err, string(stdout))
+					By(fmt.Sprintf("Killing container runtime iteration %d", i))
+					// Wait for container runtime to be running
+					var runtimePids []int
+					Eventually(func() error {
+						var err error
+						runtimePids, err = getPidsForProcess(framework.TestContext.ContainerRuntimeProcess, framework.TestContext.ContainerRuntimePidFile)
+						if err != nil {
+							return err
+						}
+						// Make sure the container runtime is running, pid got from pid file may not be running.
+						for _, pid := range runtimePids {
+							if _, err := exec.Command("sudo", "ps", "-p", fmt.Sprintf("%d", pid)).CombinedOutput(); err != nil {
+								return err
+							}
+						}
+						return nil
+					}, 1*time.Minute, 2*time.Second).Should(BeNil())
+					for _, pid := range runtimePids {
```

Comment:
When do we expect multiple pids?

Reply:
Asked the same question when I review the original container manager test PR. https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/container_manager_test.go#L80-L86

The function `getPidsForProcess` returns multiple pids. We can assume there is only one, or do things in a loop. This PR go with the second option.

Reply:
Can we check the number of pids and error out if there are more than one?

Reply:
Yeah, that's option one. We can.

Reply:
Done for restart_test.go. Leave container_manager_test.go as it is.